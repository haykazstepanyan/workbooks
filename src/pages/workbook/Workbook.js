import { useCallback } from "react";
import { useDispatch } from "react-redux";
import PropTypes from "prop-types";
import {
	deleteWorkbook,
	resetDeleteStatus,
} from "../../redux/workbook/actionCreators";
import Accordion from "../../components/accordion";
import Workplaces from "../../components/workplace";
import Table from "../../components/table";
import styles from "./workbook.module.scss";

function Workbook({ workbook, workbookID, user }) {
	const {
		firstName,
		lastName,
		user_ID,
		Workplaces: workbookWorkplaces,
		...workbookRest
	} = workbook;

	const dispatch = useDispatch();

	const tableData = {
		...workbookRest,
		birthDate: new Date(workbookRest.birthDate).toDateString().slice(4),
	};

	let canDeleteWorkbook = true;
	if (user.type === "user") {
		canDeleteWorkbook = user_ID === user.id;
	}

	const handleDeleteWorkbook = useCallback(() => {
		dispatch(deleteWorkbook(workbookID));
		setTimeout(() => {
			dispatch(resetDeleteStatus());
		}, 5000);
	}, [dispatch, workbookID]);

	return (
		<div className={styles.workbook} data-testid="workbook">
			<Accordion
				rowName={`${firstName} ${lastName}`}
				fullWidth
				canDelete={canDeleteWorkbook}
				handleDelete={handleDeleteWorkbook}
			>
				<Table tableData={tableData} />
				<Workplaces
					workbookID={workbookID}
					workplaces={workbookWorkplaces}
					workbookUserID={user_ID}
					user={user}
				/>
			</Accordion>
		</div>
	);
}

Workbook.propTypes = {
	workbook: PropTypes.shape({
		birthDate: PropTypes.string,
		email: PropTypes.string,
		passport: PropTypes.string,
		Workplaces: PropTypes.object,
	}),
	workbookID: PropTypes.string,
	user: PropTypes.objectOf(PropTypes.string),
};

export default Workbook;
