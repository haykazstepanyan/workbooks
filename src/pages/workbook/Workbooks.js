import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { getCurrentUser } from "../../redux/auth/actionCreators";
import { getWorkbooks } from "../../redux/workbook/actionCreators";
import { selectUserData } from "../../redux/auth/selectors";
import {
	selectDeleteStatus,
	selectWorkbooks,
} from "../../redux/workbook/selectors";
import Navbar from "../../components/navbar";
import { AddNewWorkbook } from "../../components/workbook";
import Workbook from "./Workbook";
import Loader from "../../components/loader";
import styles from "./workbook.module.scss";
import RequestResult from "../../components/requestResult";

function Workbooks() {
	const history = useHistory();
	const dispatch = useDispatch();
	const userData = useSelector(selectUserData);
	const workbooks = useSelector(selectWorkbooks);
	const deleteStatus = useSelector(selectDeleteStatus);

	useEffect(() => {
		dispatch(getCurrentUser());
		dispatch(getWorkbooks());
	}, [dispatch]);

	let deleteStatusText;
	if (deleteStatus === "successful") {
		deleteStatusText = "Workbook has been deleted";
	} else if (deleteStatus === "failed") {
		deleteStatusText = "Something went wrong. Please try again later";
	}

	if (!userData) {
		history.push("/login");
		return null;
	} else if (userData?.status === "pending") {
		return (
			<div className="page-center-loader">
				<Loader />
			</div>
		);
	} else {
		return (
			<>
				<Navbar userType={userData?.type} />
				<div className={styles.workbooks} data-testid="workbooks">
					<AddNewWorkbook workbooks={workbooks} user={userData} />
					<div className={styles.workbooks__box}>
						{deleteStatus && (
							<RequestResult status={deleteStatus} text={deleteStatusText} />
						)}
						{workbooks &&
							Object.entries(workbooks).map(([key, value]) => (
								<Workbook
									workbookID={key}
									workbook={value}
									user={userData}
									key={key}
								/>
							))}
					</div>
				</div>
			</>
		);
	}
}

export default Workbooks;
