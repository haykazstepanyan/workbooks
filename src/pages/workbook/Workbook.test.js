import { render } from "@testing-library/react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import Workbooks from "./";
import Workbook from "./Workbook";

describe("workbook pages components", () => {
	const mockStore = configureStore([thunk]);
	const store = mockStore({
		workbook: {},
		auth: {
			userData: {},
		},
		workplace: {},
	});

	describe("Workbooks component", () => {
		it("should match Workbooks component snapshot", () => {
			const { getByTestId } = render(
				<Provider store={store}>
					<Workbooks />
				</Provider>
			);
			const workbooks = getByTestId("workbooks");
			expect(workbooks).toMatchSnapshot();
		});
	});

	describe("Workbook component", () => {
		it("should match Workbook component snapshot", () => {
			const { getByTestId } = render(
				<Provider store={store}>
					<Workbook workbook={{}} user={{}} />
				</Provider>
			);
			const workbook = getByTestId("workbook");
			expect(workbook).toMatchSnapshot();
		});
	});
});
