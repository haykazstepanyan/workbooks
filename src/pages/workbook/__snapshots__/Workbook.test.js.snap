// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workbook pages components Workbook component should match Workbook component snapshot 1`] = `
<div
  class="workbook"
  data-testid="workbook"
>
  <div
    class="MuiPaper-root MuiAccordion-root accordion accordion__fullWidth MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-controls="panel1a-content"
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root accordion__content"
      id="panel1a-header"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          undefined undefined
        </p>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root accordion__deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            id="panel1a-content"
            role="region"
          >
            <div
              class="MuiAccordionDetails-root accordion__details"
            >
              <ul
                class="table"
                data-testid="table"
              >
                <li>
                  <span>
                    birthDate
                    :
                  </span>
                  <span>
                    lid Date
                  </span>
                </li>
              </ul>
              <div
                class=""
                data-testid="workplaces"
              >
                <div
                  class="MuiPaper-root MuiAccordion-root accordion MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
                >
                  <div
                    aria-controls="panel1a-content"
                    aria-disabled="false"
                    aria-expanded="false"
                    class="MuiButtonBase-root MuiAccordionSummary-root accordion__content"
                    id="panel1a-header"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="MuiAccordionSummary-content"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                      >
                        + Add new workplace
                      </p>
                    </div>
                    <div
                      aria-disabled="false"
                      aria-hidden="true"
                      class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
                    >
                      <span
                        class="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                          />
                        </svg>
                      </span>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </div>
                  </div>
                  <div
                    class="MuiCollapse-container MuiCollapse-hidden"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      >
                        <div
                          aria-labelledby="panel1a-header"
                          id="panel1a-content"
                          role="region"
                        >
                          <div
                            class="MuiAccordionDetails-root accordion__details"
                          >
                            <form
                              class="forms"
                            >
                              <div
                                class="inputs"
                                data-testid="wrapper-box"
                              >
                                <input
                                  class="inputs__field"
                                  data-testid="input"
                                  id="company"
                                  type="text"
                                  value=""
                                />
                                <label
                                  class="inputs__labels"
                                  data-testid="label"
                                  for="company"
                                >
                                  Company
                                </label>
                              </div>
                              <div
                                class="inputs"
                                data-testid="wrapper-box"
                              >
                                <div
                                  class="react-datepicker-wrapper"
                                >
                                  <div
                                    class="react-datepicker__input-container"
                                  >
                                    <input
                                      class="inputs__field"
                                      id="startDate"
                                      type="text"
                                      value="01/07/2021"
                                    />
                                  </div>
                                </div>
                                <label
                                  class="inputs__labels"
                                  data-testid="label"
                                  for="startDate"
                                >
                                  Start date
                                </label>
                              </div>
                              <div
                                class="inputs inputs_zeroMargin"
                                data-testid="wrapper-box"
                              >
                                <div
                                  class="react-datepicker-wrapper"
                                >
                                  <div
                                    class="react-datepicker__input-container"
                                  >
                                    <input
                                      class="inputs__field"
                                      id="endDate"
                                      type="text"
                                      value=""
                                    />
                                  </div>
                                </div>
                                <label
                                  class="inputs__labels"
                                  data-testid="label"
                                  for="endDate"
                                >
                                  End date
                                </label>
                              </div>
                              <div
                                class="inputs inputs_directionRow"
                                data-testid="wrapper-box"
                              >
                                <input
                                  class="inputs__field"
                                  data-testid="input"
                                  type="checkbox"
                                  value=""
                                />
                                <label
                                  class="inputs__labels"
                                  data-testid="label"
                                >
                                  Working till now
                                </label>
                              </div>
                              <div>
                                <div
                                  class="ReactFlagsSelect-module_flagsSelect__2pfa2 workplaces__flagSelect"
                                >
                                  <button
                                    aria-expanded="false"
                                    aria-haspopup="listbox"
                                    aria-labelledby="rfs_btn"
                                    class="ReactFlagsSelect-module_selectBtn__19wW7"
                                    id="rfs_btn"
                                    style="font-size: 16px;"
                                    type="button"
                                  >
                                    <span
                                      class="ReactFlagsSelect-module_selectValue__152eS"
                                    >
                                      Select a country
                                    </span>
                                  </button>
                                </div>
                              </div>
                              <div
                                class="buttons_rightAlign"
                              >
                                <button
                                  class="buttons"
                                  data-testid="button"
                                  disabled=""
                                  type="submit"
                                >
                                  Add
                                </button>
                              </div>
                            </form>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`workbook pages components Workbooks component should match Workbooks component snapshot 1`] = `
<div
  class="workbooks"
  data-testid="workbooks"
>
  <div>
    <div
      class="MuiPaper-root MuiAccordion-root accordion MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        aria-controls="panel1a-content"
        aria-disabled="false"
        aria-expanded="false"
        class="MuiButtonBase-root MuiAccordionSummary-root accordion__content"
        id="panel1a-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            + Add new workbook
          </p>
        </div>
        <div
          aria-disabled="false"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </div>
      <div
        class="MuiCollapse-container MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="panel1a-header"
              id="panel1a-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root accordion__details"
              >
                <form
                  class="forms"
                  data-testid="workbook-form"
                >
                  <div
                    class="inputs"
                    data-testid="wrapper-box"
                  >
                    <input
                      class="inputs__field"
                      data-testid="input"
                      id="firstName"
                      pattern="[a-zA-Z]+"
                      type="text"
                      value=""
                    />
                    <label
                      class="inputs__labels"
                      data-testid="label"
                      for="firstName"
                    >
                      First Name
                    </label>
                  </div>
                  <div
                    class="inputs"
                    data-testid="wrapper-box"
                  >
                    <input
                      class="inputs__field"
                      data-testid="input"
                      id="lastName"
                      pattern="[a-zA-Z]+"
                      type="text"
                      value=""
                    />
                    <label
                      class="inputs__labels"
                      data-testid="label"
                      for="lastName"
                    >
                      Last Name
                    </label>
                  </div>
                  <div
                    class="inputs"
                    data-testid="wrapper-box"
                  >
                    <input
                      class="inputs__field"
                      data-testid="input"
                      id="passport"
                      type="text"
                      value=""
                    />
                    <label
                      class="inputs__labels"
                      data-testid="label"
                      for="passport"
                    >
                      Passport
                    </label>
                  </div>
                  <div
                    class="inputs"
                    data-testid="wrapper-box"
                  >
                    <input
                      class="inputs__field"
                      data-testid="input"
                      id="email"
                      type="email"
                      value=""
                    />
                    <label
                      class="inputs__labels"
                      data-testid="label"
                      for="email"
                    >
                      Email
                    </label>
                  </div>
                  <div
                    class="inputs"
                    data-testid="wrapper-box"
                  >
                    <div
                      class="react-datepicker-wrapper"
                    >
                      <div
                        class="react-datepicker__input-container"
                      >
                        <input
                          class="inputs__field"
                          id="birthDate"
                          type="text"
                          value="01/07/2021"
                        />
                      </div>
                    </div>
                    <label
                      class="inputs__labels"
                      data-testid="label"
                      for="birthDate"
                    >
                      Birth date
                    </label>
                  </div>
                  <div
                    class="buttons_rightAlign"
                  >
                    <button
                      class="buttons"
                      data-testid="button"
                      disabled=""
                      type="submit"
                    >
                      Add
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="workbooks__box"
  />
</div>
`;
